core/provider/common.go

```diff
@@ -11,3 +11,19 @@ func GetRoles(pc *domain.ProviderConfig, resourceType string) ([]*domain.Role, e
 
 	return nil, ErrInvalidResourceType
 }
+
+type PermissionManager struct{}
```

Comment:
Using `PermissionManager` interface would give benefit if there are possible ways to do `GetPermission`. One of the way is the `type PermissionManager struct`.

If the logic of `GetPermissions` is the same for all providers, that is a strong sign that this does not to be interfaced, hence not needed a specific struct. This will make code simpler and reduce the dependency for all providers `e.g. BigQueryProvider does not have PermissionManager inside it`. We can just call `package.Function` (provider.GetPermission). If we do this, we don't need to mock  `GetPermission` in the test.

Reply:
yes, the only thing right now is gcloud_iam provider has different implementation on how to retrieve permissions: https://github.com/odpf/guardian/blob/069f69b9a912fe6872b7e0773a23b5345349db73/plugins/providers/gcloudiam/provider.go#L141-L144

so the possible options are to use `type PermissionManager struct` like the current one, or have `provider.GetPermissions()` being called in every provider plugins as you suggested